File: May-LeetCoding-Challange-2020/Week1/383-ransom_note.cpp
class Solution {
public:
    bool canConstruct(string ransomNote, string magazine) {
        
        /* Time Complexity: O(n)
         * Space complexity: O(n)
        */
        map<char, int> a;
        
        if(magazine.length() < ransomNote.length())
            return false;

        
        for(char b : magazine){
             a[b]++;
        }
        
    
        for(char c : ransomNote){
            if(a[c] < 1){
                return false;
            }else{
                a[c]--;
            }
        }
        
        return true;
        
    }
};

File: May-LeetCoding-Challange-2020/Week1/383-ransom-note.py
class Solution:
    # Sol 1. using Counter, Brute force
    def canConstruct(self, ransomNote: str, magazine: str) -> bool:
        # Edge case: if len(ransomNote ) > len(magazine): False
        ran_col = collections.Counter(ransomNote)
        mag_col = collections.Counter(magazine)
        return ran_col&mag_col==ran_col
        # Time Complexity: O(n)
        # Space Complexity: O(n)